src/workerd/io/actor-cache-test.c++

```diff
@@ -1444,76 +1424,131 @@ KJ_TEST("ActorCache read retry on flush containing only puts") {
   KJ_ASSERT(KJ_ASSERT_NONNULL(promise.wait(ws)) == "123");
 }
 
-KJ_TEST("ActorCache read hard fail") {
+// KJ_TEST("ActorCache read hard fail") {
```

Comment:
This test is currently commented out because:
1. When we `kj::mv(mockGet).thenThrow(KJ_EXCEPTION(FAILED, "read failed"));` the read promise, we cause the `syncGet()` storage operation to throw.
2. This results in `startFlushTransaction()` to throw, which causes the `flushProm` in `flushImpl()` to throw.
3. We end up in the `catch_()` handler of the `flushProm`, and since our exception cannot be retried, we go into the last `else` and throw another exception and break the output gate.

Prior to the coalesced read implementation, we would expect the failure of a single storage get() to propagate the exception to the JS caller. This was easy because storage gets() were completely independent operations from writes. Now that `get()` goes through the same flush as writes, we've got a few more things to worry about.

---
### Propagating the Exception to the caller
We'll focus on the single key `get()` for now. [Currently](https://github.com/cloudflare/workerd/pull/1916/commits/516c7ce7fe28e264f494f9c9c9b62938d3547bb6#diff-52b552cc9d6477516b49e970aa23de236c7a1f7327b09c243c336445cbb2e18eR439-R444), the JS get caller must wait for the flush to complete, and upon success, the caller will receive the result of the `get()`.

This breaks when we force a storage `get()` to throw, because instead of succeeding, the flush will fail with `broken.outputGateBroken; jsg.Error: Internal error in Durable Object storage write caused object to be reset`. 

I believe we can fix this by waiting on a joint promise in `ActorCache::get()`. We can keep the successful flush case (linked above), and add another promise for when the storage get (in `syncGet()`) fails. To make things explicit, we can wrap much of `syncGet()` in a `try...catch`, and upon entering the catch branch, propagate the exception to the waiting `ActorCache::get()`. This would effectively ensure that the JS caller receives the exception. 

I've already tested this and confirmed it works, and it doesn't look _too_ daunting. In short, the constructor for `GetWaiter` takes all the entries we care about for this `get()`. For each `Entry`, we create a PaF, give the fulfiller to the `Entry`, and add the promise to a single joint "_the read failed_" promise owned by the `GetWaiter`.

- At this point, the `GetWaiter` has a promise that resolves when any of the associated `Entry`s use their fulfiller, indicating the entire get has failed.

Back in the `try...catch` for `syncGet()`, if we're in the catch branch, we can just use the fulfiller on the `Entry` to signal that the storage `get()` failed.

This approach should extend to the multi-key get() fairly easily, since we can just iterate over all the `Entry`s in a `GetFlush` and fulfill each of their fulfillers. **That said, I don't know if this is what we want** -- I would have to look into it more. Not sure if we have a precedent for this.
### What to do with the flush
Assuming we've propagated the exception to `ActorCache::get()`, we now need to decide what to do about this ongoing flush. Should we throw a fatal exception from `syncGet()`? Maybe only if it's a `DISCONNECTED` exception?

At least in the case of `FAILED`, if we still throw the exception, then this `ActorCache read hard fail` test will still fail because the output gate will break. Perhaps there are certain types of exceptions that should be propagated to the get() caller (`FAILED/OVERLOADED`), and certain types that should only be thrown from `syncGet()` (`DISCONNECTED` so we retry).
### Do Entry's need to be removed from dirtyList/cache?
If we don't propagate `FAILED` exceptions from `syncGet()`, do we need to worry about the `Entry`s associated with the failed get() in the `dirtyList`? 

For example, let's say we fail the storage get(), send the exception to `ActorCache::get()`, and then don't throw from `syncGet()`. Assuming there are no other storage operations, we will continue down our flush path to the successful flush case, and iterate over the `dirtyList`, moving values into the `cleanList`. This seems wrong, although perhaps if the `GetWaiter` is destroyed these entries will also be removed from the `dirtyList` before the `flushProm` continuation runs?

I would think we would need to remove the entries from `dirtyList` inside `syncGet()`, but may need to think about this some more.

Reply:
This all feels like the natural consequences of reusing `lastFlush` and the `dirtyList`. Can we keep the storage reads segregated onto a separate promise rather than integrating them so tightly into `lastFlush`?  e.g. a `coalescedReads` or `lastBatchedGet` promise or something.

It seems like that would simplify some of this desire to treat read failures separately from write failures (and, as asked about in my other comment, it'd allow reads to resolve prior to needing to wait on writes to flush).

---

As for removing entries from the `dirtyList`, I'm not sure I totally understand the question unfortunately. If we're throwing an exception to the caller for a given `get`, then I'd think we should remove it from the list (although presumably not by adding it to the `cleanList` given that we don't know its value). If we're still going to try to get the key's value and return it, then we shouldn't. But I suspect that isn't an answer to your question.